src/fiber.cr

```diff
@@ -133,6 +118,17 @@ class Fiber
     end
   end
 
+  # :nodoc:
+  def stack : Slice(Void)
+    # If we allocated our own stack, we know its bounds.
+    if @stack != Pointer(Void).null
+      return Slice(Void).new(@stack, @stack_bottom.address - @stack.address)
+    end
+    # Otherwise, assume for now that there is only one thread,
```

Comment:
Maybe use a more detailed explanation:
> the fiber represents a thread's main stack and we assume it's the currently running thread; this should be fine since a thread should never resume the main fiber of another thread (undefined behavior? chaos?)

Reply:
Why not pass the thread stack through the `Fiber.new` constructor. Then we simply have a fiber constructor for creating a fiber with a new stack and a fiber constructor for creating a fiber with an existing stack, and `Fiber` never has to worry about `Thread.current.stack` at all.

Reply:
We can’t cache @stack in Fiber because it is dynamic in time. What is is at the time of Fibre construction need not be what it is when a segfault occurs later in the run. The stack bottom stays put but the stack size adjusts.

Reply:
I did too much low level programming, I guess. It sounded safe enough, but it's better to be really safe.

Is the stack size really changing? We never tweak it, and I'm not sure we ever do.

If so, maybe linking a Fiber to its Thread when it represents a thread stack would be a better idea? We could then have something like `@thread.try(&.stack) || @stack`.

Reply:
On `x86_64-linux-musl` at least, the stack sizes reported from `pthread_attr_get_np` & `pthread_attr_getstack` on the main thread reflect the current stack allocation, which can be much smaller than the ultimate limit, so it will grow in time. And although the Crystal code base never changes the limits on the main stack, there is nothing to stop a user from calling `setrlimit`, which will change the main stack size.

Saving the associated thread if the Fiber represents a thread's stack (main thread or otherwise), is the right thing to do.

Reply:
Well, it turned out that it having the `Thread` know about its `Fiber` and the `Fiber` knowing about its `Thread` didn't mix well with the fact that `Thread` has a finalizer.  Numerous `GC Warning: Finalization cycle` warnings were generated running the spec suite.

So instead, a `Fiber` returns an optional `Thread::Stack` which is `Nil` if the `Fiber` is not responsible for maintaining its stack (i.e. the main `Fiber` for a `Thread`).  A `Thread` always knows about its own stack and can return it.  If you want the current stack, talk to the `Crystal::Scheduler`, which knows both the current `Thread` and the current `Fiber` and can figure out what is the right stack to report.